### PR TITLE
Add "Auto Rotate" orientation for wide pages

### LINKED ISSUE #1977: Automatically rotate based on image size/aspect ratio
**App version:**
0.8.4

**Issue/Request:**

Change screen orientation by image size

**Other details:**
If the app detects a landscape image, like 4:3 or 16:9, the whole app itself automatically rotate to landscape orientation so there's no need to wait/reply on the sensor. This also benefits users who read manga lying down sideways(Thereby they can't rely on the sensor)

This feature is also present under [ QuickPic app > Settings > View > Change screen orientation > By image size ] for reference.

### PR DESCRIPTION
**Pull request by timothyng-164**
Closes #1977

This addresses [#1977](https://github.com/tachiyomiorg/tachiyomi/issues/1977) by adding the "Dual page rotate" option to the paged reader.

Feel free to edit and merge this branch as needed.

**New Reader Settings:**
- Dual page rotate
- Reverse dual page rotate orientation (only shown if the former is checked)

<a href="url"></a>
<hr></hr>

**Behavior for Dual Pages**:

1. Default Behavior for Dual Pages
2. "Dual page rotate" checked
3. "Reverse dual page rotate orientation" checked

<a href="url"></a>

<a href="url"></a>

<a href="url"></a>

### CODE REVIEW

**Comment by jferdous70** [comment]
could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.

**Comment by timothyng-164** [comment]
> could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.

Sure, I'm thinking there could be 2 orientations:
- Auto Rotate Landscape - landscape on wide image, portrait on thin
- Auto Rotate Portrait - landscape on thin image, portrait on wide

I'm not sure if there's a better name but would that work?

**Comment by jferdous70** [comment]
> > could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.
> 
> Sure, I'm thinking there could be 2 orientations:
> 
>     * Auto Rotate Landscape - landscape on wide image, portrait on thin
> 
>     * Auto Rotate Portrait - landscape on thin image, portrait on wide
> 
> 
> I'm not sure if there's a better name but would that work?

That's not exactly what I wanted, but I now realized main tachiyomi don't about webtoon detection feature

**Comment by timothyng-164** [comment]
> > > could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.
> > 
> > 
> > Sure, I'm thinking there could be 2 orientations:
> > ```
> > * Auto Rotate Landscape - landscape on wide image, portrait on thin
> > 
> > * Auto Rotate Portrait - landscape on thin image, portrait on wide
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm not sure if there's a better name but would that work?
> 
> That's not exactly what I wanted, but I now realized main tachiyomi don't about webtoon detection feature

I see, I reverted th

> > > could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.
> > 
> > 
> > Sure, I'm thinking there could be 2 orientations:
> > ```
> > * Auto Rotate Landscape - landscape on wide image, portrait on thin
> > 
> > * Auto Rotate Portrait - landscape on thin image, portrait on wide
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm not sure if there's a better name but would that work?
> 
> That's not exactly what I wanted, but I now realized main tachiyomi don't about webtoon detection feature

I see, I reverted the change to just have the "Auto Rotate" orientation".

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderActivity.kt [hunk 1]
```diff
@@ -778,6 +780,18 @@ class ReaderActivity : BaseRxActivity<ReaderPresenter>() {
...
         binding.pageSlider.value = page.index.toFloat()
+
+        if (presenter.getMangaOrientationType() == OrientationType.AUTO_ROTATE.flagValue) {
+            rotateOnWideImage(page)
+        }
+    }
+
+    private fun rotateOnWideImage(page: ReaderPage) {
+        val stream = page.stream!!
+        when (ImageUtil.isWideImage(stream.invoke())) {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderActivity.kt:791]: You should only really need to read the header, similar to usages of the existing `isWideImage` function that takes in a buffered stream.

File: app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt [hunk 2]
```diff
@@ -555,7 +565,7 @@ object ImageUtil {
     /**
      * Used to check an image's dimensions without loading it in the memory.
      */
-    private fun extractImageOptions(
+    fun extractImageOptions(
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt:568]: Can stay private?

Comment (timothyng-164) [reply app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt]: This has been reverted

File: app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderActivity.kt (hunk 1 above)
Comment (timothyng-164) [reply app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderActivity.kt]: I believe this has been addressed

**Review by arkon — COMMENTED** [review]
General feedback:
- In its current implementation, I feel like this should only happen for paged reading modes. It's pretty janky to suddenly rotate in continuous vertical or webtoon modes.
- I'm wondering if we should just be loading a rotated version of the image rather than actually changing the system orientation, which requires a full activity recreation?

**Comment by timothyng-164** [comment]
Good point, I didn't consider continuous vertical or webtoons. 
I like the idea of just loading the image already rotated. I added a couple options to the reader settings.

**Comment by arkon** [comment]
Rebased/cleaned up a bit in 953720472fe64ef488ecae7ae7fea453b8c7c68b

**Pull request closed without merging**